### PR TITLE
Promoting ogr_layer_algebra to official python script

### PR DESCRIPTION
**Pull request by kssvrk**
## What does this PR do?

This PR  Promotes ogr_layer_algebra.py to official Python script mentioned at issue [1581](https://github.com/OSGeo/gdal/issues/1581). 

What i have done is the following.

- move ogr_layer_algebra.py from swig/python/gdal-utils/osgeo_utils/samples to swig/python/gdal-utils/osgeo_utils/
- create a calling gdal-utils script for ogr_layer_algebra.py in swig/python/gdal-utils/scripts
- Change the arguments of main function of ogr_layer_algebra from argv=None to argv=sys.argv
- created documentation for ogr_layer_algebra at doc/source/programs/
- Linked the documentation for ogr_layer_algebra in index under vector section
- Created test cases for all modes of ogr_layer_algebra at autotest/pyscripts/test_ogr_layer_algebra.py

## What are related issues/pull requests?

Tries to solve Issue [1581](https://github.com/OSGeo/gdal/issues/1581). 

## Tasklist

-  [x] Reorganised the script 
 - [x] Add test case
 - [x] Add documentation

### CODE REVIEW

**Comment by rouault** [comment]
(please don't close the pull request each time continuous integration raises error. you can just add new commits to fix issues and push again your branch)

could you do the following documented in https://github.com/OSGeo/gdal/blob/master/CONTRIBUTING.md#build-gdal-cmake:
Setup pre-commit (to be done just once):
```
python -m pip install pre-commit
pre-commit install
```
and then you can automatically fix the .py files with: ``pre-commit --run-all -a`` (this is also automatically done on modified un-committed files when running git commit)

**Review by kssvrk — COMMENTED** [review]
@rouault I have done the required changes with pre-commit hook.
black,flake8,isort have passed.
pytest is also succesful for the test script written.
Please let me know if i have to look into anything else.

File: swig/python/gdal-utils/scripts/ogr_layer_algebra.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env python3
+
+import sys
+
+from osgeo.gdal import deprecation_warn
+
+# import osgeo_utils.ogrmerge as a convenience to use as a script
```
Comment (rouault) [inline swig/python/gdal-utils/scripts/ogr_layer_algebra.py:7]:
```suggestion
# import osgeo_utils.ogr_layer_algebra as a convenience to use as a script
```

File: autotest/pyscripts/test_ogr_layer_algebra.py [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+import test_py_scripts
+
+from osgeo import ogr
+
+###############################################################################
+
+# Test Intersection
+
+
+def test_ogr_layer_algebra_1():
```
Comment (rouault) [inline autotest/pyscripts/test_ogr_layer_algebra.py:41]:
```suggestion
def test_ogr_layer_algebra_intersection():
```

File: autotest/pyscripts/test_ogr_layer_algebra.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+
+    assert featureCount == 2
+
+
+###############################################################################
+
+# Test Union
+
+
+def test_ogr_layer_algebra_2():
```
Comment (rouault) [inline autotest/pyscripts/test_ogr_layer_algebra.py:104]:
```suggestion
def test_ogr_layer_algebra_union():
```

File: autotest/pyscripts/test_ogr_layer_algebra.py [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+
+    assert featureCount == 5
+
+
+###############################################################################
+
+# Test Symmetric Difference
+
+
+def test_ogr_layer_algebra_3():
```
Comment (rouault) [inline autotest/pyscripts/test_ogr_layer_algebra.py:167]:
```suggestion
def test_ogr_layer_algebra_symdifference():
```

File: autotest/pyscripts/test_ogr_layer_algebra.py [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+
+    assert featureCount == 4
+
+
+###############################################################################
+
+# Test Identity
+
+
+def test_ogr_layer_algebra_4():
```
Comment (rouault) [inline autotest/pyscripts/test_ogr_layer_algebra.py:235]:
```suggestion
def test_ogr_layer_algebra_identity():
```

File: autotest/pyscripts/test_ogr_layer_algebra.py [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+
+    assert featureCount == 4
+
+
+###############################################################################
+
+# Test Update
+
+
+def test_ogr_layer_algebra_5():
```
Comment (rouault) [inline autotest/pyscripts/test_ogr_layer_algebra.py:303]:
```suggestion
def test_ogr_layer_algebra_update():
```

File: autotest/pyscripts/test_ogr_layer_algebra.py [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+
+    assert featureCount == 3
+
+
+###############################################################################
+
+# Test Clip
+
+
+def test_ogr_layer_algebra_6():
```
Comment (rouault) [inline autotest/pyscripts/test_ogr_layer_algebra.py:366]:
```suggestion
def test_ogr_layer_algebra_clip():
```

File: autotest/pyscripts/test_ogr_layer_algebra.py [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+
+    assert featureCount == 2
+
+
+###############################################################################
+
+# Test Erase
+
+
+def test_ogr_layer_algebra_7():
```
Comment (rouault) [inline autotest/pyscripts/test_ogr_layer_algebra.py:429]:
```suggestion
def test_ogr_layer_algebra_erase():
```

File: doc/source/programs/ogr_layer_algebra.rst [hunk 9]
```diff
@@ -0,0 +1,169 @@
+.. _ogr_layer_algebra:
+
+================================================================================
+ogr_layer_algebra.py
+================================================================================
+
```
Comment (rouault) [inline doc/source/programs/ogr_layer_algebra.rst:6]:
```suggestion

.. versionadded:: 3.6

```

File: doc/source/programs/ogr_layer_algebra.rst [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    Indicates wether the ``output_ds`` have to be overwritten with the generated result of ogr_layer_algebra
+
+.. option:: -opt <NAME=VALUE>
+
+    Attributes for which the operation has to run on ``input_ds`` and ``method_ds``
+
+.. option:: -f <format_name>
+
+    Select the output format. Starting with GDAL 2.3, if not specified,
```
Comment (rouault) [inline doc/source/programs/ogr_layer_algebra.rst:108]:
```suggestion
    Select the output format. If not specified,
```

File: doc/source/programs/ogr_layer_algebra.rst [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    a non-linear type to its generalized curve type (POLYGON to CURVEPOLYGON,
+    MULTIPOLYGON to MULTISURFACE, LINESTRING to COMPOUNDCURVE, MULTILINESTRING 
+    to MULTICURVE). Starting with version 2.1 the type can be defined as measured 
+    (“25D” remains as an alias for single “Z”). Some forced geometry conversions 
+    may result in invalid geometries, for example when forcing conversion 
+    of multi-part multipolygons with -nlt POLYGON, the resulting polygon will
+    break the Simple Features rules.
+
+    Starting with GDAL 3.0.5, -nlt CONVERT_TO_LINEAR and 
+    -nlt PROMOTE_TO_MULTI can be used simultaneously.
```
Comment (rouault) [inline doc/source/programs/ogr_layer_algebra.rst:153]:
```suggestion
    One of NONE, GEOMETRY, POINT, LINESTRING, POLYGON, GEOMETRYCOLLECTION,
    MULTIPOINT, MULTIPOLYGON, GEOMETRY25D, POINT25D, LINESTRING25D, POLYGON25D,
    GEOMETRYCOLLECTION25D, MULTIPOINT25D, MULTIPOLYGON25D.
```

File: doc/source/programs/ogr_layer_algebra.rst [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+.. option:: -a_srs <srs_def>
+
+    Assign an output SRS, but without reprojecting (use -t_srs to reproject)
```
Comment (rouault) [inline doc/source/programs/ogr_layer_algebra.rst:157]:
```suggestion
    Assign an output SRS, but without reprojecting
```

File: autotest/pyscripts/test_ogr_layer_algebra.py [hunk 13]
```diff
@@ -0,0 +1,489 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+###############################################################################
+
+import pytest
+import test_py_scripts
+
+from osgeo import ogr
+
```
Comment (rouault) [inline autotest/pyscripts/test_ogr_layer_algebra.py:35]: as ogr_layer_algebra algorithms require GEOS support, we must skip the tests if it is not available:

```suggestion

import ogrtest

###############################################################################
@pytest.fixture(autouse=True, scope="module")
def startup_and_cleanup():

    if not ogrtest.have_geos():
        pytest.skip()


```

File: autotest/pyscripts/test_ogr_layer_algebra.py (hunk 2 above)
Comment (kssvrk) [reply autotest/pyscripts/test_ogr_layer_algebra.py]: I thought of naming the functions in the way you prescribed, but just followed other tests thinking that it was a general convention.

Now changed as you suggested in https://github.com/kssvrk/gdal/commit/841f3fe4b409219623527f0a021c34eae6071697

File: autotest/pyscripts/test_ogr_layer_algebra.py (hunk 3 above)
Comment (kssvrk) [reply autotest/pyscripts/test_ogr_layer_algebra.py]: Changed in https://github.com/kssvrk/gdal/commit/841f3fe4b409219623527f0a021c34eae6071697

File: autotest/pyscripts/test_ogr_layer_algebra.py (hunk 4 above)
Comment (kssvrk) [reply autotest/pyscripts/test_ogr_layer_algebra.py]: Changed in https://github.com/kssvrk/gdal/commit/841f3fe4b409219623527f0a021c34eae6071697

File: autotest/pyscripts/test_ogr_layer_algebra.py (hunk 5 above)
Comment (kssvrk) [reply autotest/pyscripts/test_ogr_layer_algebra.py]: Changed in https://github.com/kssvrk/gdal/commit/841f3fe4b409219623527f0a021c34eae6071697

File: autotest/pyscripts/test_ogr_layer_algebra.py (hunk 6 above)
Comment (kssvrk) [reply autotest/pyscripts/test_ogr_layer_algebra.py]: Changed in https://github.com/kssvrk/gdal/commit/841f3fe4b409219623527f0a021c34eae6071697

File: autotest/pyscripts/test_ogr_layer_algebra.py (hunk 7 above)
Comment (kssvrk) [reply autotest/pyscripts/test_ogr_layer_algebra.py]: Changed in https://github.com/kssvrk/gdal/commit/841f3fe4b409219623527f0a021c34eae6071697

File: autotest/pyscripts/test_ogr_layer_algebra.py (hunk 8 above)
Comment (kssvrk) [reply autotest/pyscripts/test_ogr_layer_algebra.py]: Changed in https://github.com/kssvrk/gdal/commit/841f3fe4b409219623527f0a021c34eae6071697

File: doc/source/programs/ogr_layer_algebra.rst (hunk 9 above)
Comment (kssvrk) [reply doc/source/programs/ogr_layer_algebra.rst:6]: Added this in https://github.com/kssvrk/gdal/commit/0a8393cee671b57f498c051b3ba83a1c5401f6c4

File: doc/source/programs/ogr_layer_algebra.rst (hunk 10 above)
Comment (kssvrk) [reply doc/source/programs/ogr_layer_algebra.rst]: Added this in https://github.com/kssvrk/gdal/commit/0a8393cee671b57f498c051b3ba83a1c5401f6c4

File: doc/source/programs/ogr_layer_algebra.rst (hunk 11 above)
Comment (kssvrk) [reply doc/source/programs/ogr_layer_algebra.rst]: Added this in https://github.com/kssvrk/gdal/commit/0a8393cee671b57f498c051b3ba83a1c5401f6c4

File: doc/source/programs/ogr_layer_algebra.rst (hunk 12 above)
Comment (kssvrk) [reply doc/source/programs/ogr_layer_algebra.rst]: Added this in https://github.com/kssvrk/gdal/commit/0a8393cee671b57f498c051b3ba83a1c5401f6c4

File: autotest/pyscripts/test_ogr_layer_algebra.py (hunk 13 above)
Comment (kssvrk) [reply autotest/pyscripts/test_ogr_layer_algebra.py:36]: changed in https://github.com/kssvrk/gdal/commit/841f3fe4b409219623527f0a021c34eae6071697

File: swig/python/gdal-utils/scripts/ogr_layer_algebra.py (hunk 1 above)
Comment (kssvrk) [reply swig/python/gdal-utils/scripts/ogr_layer_algebra.py]: added in https://github.com/kssvrk/gdal/commit/6da6052853cf70eec7e4bb703873e12c037f7b07

**Comment by kssvrk** [comment]
@rouault I have done the recommended changes by you. Kindly review and provide feedback.

File: doc/source/programs/ogr_layer_algebra.rst [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+    * ``Erase``
+
+        The erase method creates a layer, which has features from the input layer whose areas are erased by the features in the method layer. 
+        By default the result layer has attributes of the input layer. 
+        
+.. option:: -input_ds <path>
+
+    Input data set path for the operation to be performed.
+     For operations involving 2 datasets, this is one of the dataset.
```
Comment (rouault) [inline doc/source/programs/ogr_layer_algebra.rst:76]:
```suggestion
    For operations involving 2 datasets, this is one of the dataset.
```

**Pull request merged**